### PR TITLE
docs: add reactor sections

### PR DESCRIPTION
**Pull request by marbar3778**
## Description 

Add documentation reactor specs

### CODE REVIEW

File: docs/tendermint-core/reactors/block-sync/bcv0/README.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+order: 1
+parent:
+  title: v0
+  order: 1
+---
+
+# BCV0
```
Comment (marbar3778) [inline docs/tendermint-core/reactors/block-sync/bcv0/README.md:8]: add more info here

File: docs/tendermint-core/reactors/README.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  title: Reactors
+  order: 1
+---
+
+# Reactors
+
+Tendermint-Core is split into many components, some of these components interact with other nodes in a network. These components are called `Reactors`. Each Reactor has its own functionality, we will be walking through the reactors within Tendermint-Core.
+
+
+- [Consensus]()
```
Comment (marbar3778) [inline docs/tendermint-core/reactors/README.md:13]: add links

File: docs/tendermint-core/reactors/evidence/README.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+parent:
+  title: Mempool
+  order: 3
+---
+
+# Evidence
+
+Evidence is used to identify validators who have or are acting malicious. There are multiple types of evidence, to read more on the evidence types please see [Evidence Types](https://docs.tendermint.com/master/spec/core/data_structures.html#evidence).
+
+The evidence works
```
Comment (marbar3778) [inline docs/tendermint-core/reactors/evidence/README.md:12]: finish or delete

File: docs/tendermint-core/README.md [hunk 4]
```diff
@@ -18,3 +20,4 @@ This section dives into the internals of Go-Tendermint.
 - [State Sync](./state-sync.md)
 - [Mempool](./mempool.md)
 - [Light Client](./light-client.md)
+- [Reactors](./reactors/README.md)
```
Comment (cmwaters) [inline docs/tendermint-core/README.md:23]: Do you think we should have a reactors section if we already have a `Mempool`, `StateSync`, `FastSync` sections. I would say we just merge them with the existing ones if the are there.

And then we just need to add `Consensus`, `Evidence` and `PeerExchange`

Comment (tychoish) [reply docs/tendermint-core/README.md:23]: If we make all of these private/internal, maybe we want to think about how we document them?

File: docs/tendermint-core/reactors/block-sync/bcv0/implementation.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
+---
+order: 3
+---
+
+# Implementation
```
Comment (tychoish) [inline docs/tendermint-core/reactors/block-sync/bcv0/implementation.md:5]: there are many things in this file that ought to be marked as code/pre with backtics

File: docs/tendermint-core/README.md (hunk 4 above)
Comment (marbar3778) [reply docs/tendermint-core/README.md:23]: The docs here are meant more as a implementation spec, this is orthogonal to internal.

**Comment by milosevic** [comment]
> I was talking to Marko about documentation and I wanted to share some of my thoughts here (perhaps there's even a better place to have a discussion but here will do for now). When we are thinking about how to structure our documentation, I believe we should map it to the kinds of people that will want to use the tendermint docs. There are five main kinds:
> 
> 1. Node Operators / Validators who want to know how to set up a node, how to upgrade, set up a remote signer, debug problems etc...
> 2. App Developers who want to know how to use ABCI and build their application on top of Tendermint, how to run test nets.
> 3. External Client Developers like wallets and explorers who want to know how to connect and query the RPC
> 4. Enthusiasts who want a non-code high level description of a few of the core parts of tendermint i.e. consensus and mempool.
> 5. Possible contributors who want to understand the internals, how the reactors work, the p2p layer and so forth.
>    (there may be some overlap over the kinds of people)
> 
> IMO, the spec shouldn't be included but rather we refer to it in certain places of section 5 where we feel necessary. In reality the spec predominantly serves us as the team (and any other team that wants to write an implementation of Tendermint but I don't think that happens too often)

I second this, and I would just make one more distinction re 4 and 5: we probably want to make difference between technical documentation that are used by developers to document the implementation and more high level protocol specification where high level ideas and properties are captured. The latter is normally used more directly as an artifact for audit/verification efforts, while the former is something primarily used by the implementation team. We were not necessarily doing great job making this distinction, so different specifications fit in one of these two categories, primarily depending on who has written the original version. it might be good idea in future being more explicit about intended audience as it turns out that it is very hard writing docs for multiple audiences. However, we probably want to have links between docs and ideally keeping them in sync.

File: docs/tendermint-core/block-sync/README.md [hunk 6]
```diff
@@ -1,7 +1,11 @@
 ---
-order: 10
+order: 1
+parent:
+  title: Fast Sync
```
Comment (cmwaters) [inline docs/tendermint-core/block-sync/README.md:4]:
```suggestion
  title: Block Sync
```
?

File: docs/tendermint-core/consensus/README.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+only to a subset of processes called peers. By the gossiping protocol, a validator send to its peers
+all needed information (`ProposalMessage`, `VoteMessage` and `BlockPartMessage`) so they can
+reach agreement on some block, and also obtain the content of the chosen block (block parts). As
+part of the gossiping protocol, processes also send auxiliary messages that inform peers about the
+executed steps of the core consensus algorithm (`NewRoundStepMessage` and `NewValidBlockMessage`), and
+also messages that inform peers what votes the process has seen (`HasVoteMessage`,
+`VoteSetMaj23Message` and `VoteSetBitsMessage`). These messages are then used in the gossiping
+protocol to determine what messages a process should send to its peers.
+
+We now describe the content of each message exchanged during Tendermint consensus protocol.
```
Comment (cmwaters) [inline docs/tendermint-core/consensus/README.md:41]: It seems like this is unfinished?

File: docs/tendermint-core/consensus/README.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
+---
+order: 1
+parent:
+  title: Consensus
+  order: 6
+---
+
+
+Tendermint Consensus is a distributed protocol executed by validator processes to agree on
```
Comment (cmwaters) [inline docs/tendermint-core/consensus/README.md:9]: Would be great to add a reference to the white paper somewhere

File: docs/tendermint-core/evidence/README.md [hunk 9]
```diff
@@ -0,0 +1,13 @@
+---
+order: 1
+parent:
+  title: Mempool
```
Comment (cmwaters) [inline docs/tendermint-core/evidence/README.md:4]:
```suggestion
  title: Evidence
```

File: docs/tendermint-core/pex/README.md [hunk 10]
```diff
@@ -0,0 +1,177 @@
+---
+order: 1
+parent:
+  title: Peer Exchange
+  order: 5
+---
+
+# Peer Strategy and Exchange
+
+Here we outline the design of the AddressBook
```
Comment (cmwaters) [inline docs/tendermint-core/pex/README.md:10]: I believe now it is called a peer store

**Review by cmwaters — COMMENTED** [review]
Would be great to get all this work over the line.

Q: We have `doc.go` files and then the `docs/tendermint-core/**reactor**/README.md` which seem to have a large degree of overlap. Do you think it makes sense to converge on what document location for reactors

File: docs/tendermint-core/consensus/README.md (hunk 7 above)
Comment (cmwaters) [reply docs/tendermint-core/consensus/README.md:41]: Or is this meant to lead us on to the next pages?

**Comment by marbar3778** [comment]
> Q: We have `doc.go` files and then the `docs/tendermint-core/**reactor**/README.md` which seem to have a large degree of overlap. Do you think it makes sense to converge on what document location for reactors

we can copy over, but generally doc.go may share the same intent as the docs. This is what i have seen with open source go code

**Pull request closed without merging**